### PR TITLE
feat(frontend): ic token zod

### PR DESCRIPTION
**Pull request by peterpeterparker**
# Motivation

Convert ic token related definition to zod.

# Changes

- Extract schema validation and tests
- Convert types
- Reorder types to avoid having declaration referencing types declared afterwards

### CODE REVIEW

File: src/frontend/src/lib/types/token.ts [hunk 1]
```diff
@@ -40,7 +40,7 @@ const TokenBuyableSchema = z.object({
 	buy: z.custom<AtLeastOne<TokenBuy>>().optional()
 });
 
-const TokenSchema = z
+export const TokenSchema = z
```
Comment (peterpeterparker) [inline src/frontend/src/lib/types/token.ts:43]: I did not provided a test for this. I guess we should move it to a `token.validation` in another PR and add a test. I can do it before this PR if you wish?

Comment (peterpeterparker) [reply src/frontend/src/lib/types/token.ts:43]: https://github.com/dfinity/oisy-wallet/pull/3340

**Review by AntonioVentilii — COMMENTED** [review]
LGTM tks!

One note: when I did it for the `Token` type, there were a few sub-types that required only the schema, and not the `infer type of...`. We currently use all of them? Apart for the tests I mean

**Review by AntonioVentilii — APPROVED** [review]
LGTM tks!

One note: when I did it for the Token type, there were a few sub-types that required only the schema, and not the infer type of.... We currently use all of them? Apart for the tests I mean

**Comment by peterpeterparker** [comment]
> We currently use all of them?

Only one was not use, `IcAppMetadata`. I prefered to not removed it here since it was not really the goal of the PR.

**Comment by peterpeterparker** [comment]
Relates to https://github.com/dfinity/oisy-wallet/issues/2883

**Pull request merged**